modules/consensus.go

```diff
@@ -30,9 +30,22 @@ type (
 	ConsensusSetSubscriber interface {
 		// ProcessConsensusChange sends a consensus update to a module through
 		// a function call. Updates will always be sent in the correct order.
+		// There may not be any reverted blocks, but there will always be
+		// applied blocks.
 		ProcessConsensusChange(ConsensusChange)
 	}
 
+	// ConsensusSetDigestSubscriber receives digests about changes to the
+	// consensus set in the form of a list of ids of blocks that got reverted
+	// and a list of ids of blocks that got applied. The ordering is always all
+	// of the reverted blocks followed by all of the applied blocks.
+	ConsensusSetDigestSubscriber interface {
+		// ProcessConsensusDigest sends a list of reverted blocks and applied
+		// blocks to the module. There may not be any reverted blocks, but
+		// there will always be applied blocks.
+		ProcessConsensusDigest(revertedBlocks []types.BlockID, appliedBlocks []types.BlockID)
```

Comment:
Not any more than is already happening, at least not at startup. Live, a block may be decoded 5 or more times as opposed to just once, but we're talking about 30 milliseconds, which is not a problem since validation takes much longer than that anyway.

I'm focusing down the biggest bottlenecks for now.